filedescriptor/src/unix.rs

```diff
@@ -217,6 +265,41 @@ impl FileDescriptor {
         }
         Ok(())
     }
+
+    /// Attempt to duplicate the underlying handle from an object that is
+    /// representable as the system `RawFileDescriptor` type and assign it to
+    /// a destination file descriptor. It then returns a `FileDescriptor`
+    /// wrapped around the duplicate.  Since the duplication requires kernel
+    /// resources that may not be available, this is a potentially fallible operation.
+    /// The returned handle has a separate lifetime from the source, but
+    /// references the same object at the kernel level.
+    pub unsafe fn dup2<F: AsRawFileDescriptor>(f: &F, dest_fd: RawFd) -> anyhow::Result<Self> {
+        OwnedHandle::dup2_impl(f, dest_fd).map(|handle| Self { handle })
+    }
+
+    pub(crate) fn redirect_stdin<F: AsRawFileDescriptor>(f: &F) -> anyhow::Result<FileDescriptor> {
+        let std_original = FileDescriptor::dup(libc::STDIN_FILENO)?;
+        let cloned_handle = OwnedHandle::dup(f)?;
+        unsafe { FileDescriptor::dup2(cloned_handle.handle, libc::STDIN_FILENO) }?;
+
+        Ok(std_original)
+    }
+
+    pub(crate) fn redirect_stdout<F: AsRawFileDescriptor>(f: &F) -> anyhow::Result<FileDescriptor> {
```

Comment:
FWIW, as a matter of style, I was envisioning this being a single function like this:

```rust
pub(crate) fn redirect_stdio_impl<F: AsRawFileDescriptor>(
        f: &F,
        stdio: StdioDescriptor,
    ) -> anyhow::Result<Self> {
```

that internally did:

```rust
let std_original = FileDescriptor::dup(match stdio {
   StdioDescriptor::Stdin => libc::STDIN_FILENO,
   StdioDescriptor::Stdout => libc::STDOUT_FILENO,
   StdioDescriptor::Stderr => libc::STDERR_FILENO,
))?;
```

and similarly for the Windows implementation

Reply:
Ah ok sure. Just thought the execution would be slightly faster with what I've done currently as it avoids having to move the enum into the interior methods. Will change this though to do it this way instead then 👍

Reply:
FWIW, that enum is super cheap to move around because it fits in a register.  Also: the runtime cost of this code path will be dominated by the cost of the file descriptor related syscalls, so even if the enum was big it still wouldn't matter.

Reply:
Cool, have now made the change